tidy3d/web/api/autograd/autograd.py

```diff
@@ -862,6 +862,29 @@ def postprocess_adj(
         if structure.background_medium is not None:
             eps_out = structure.background_medium.eps_model(freq_adj)
 
+        if isinstance(structure.geometry, (td.PolySlab, td.Cylinder)):
+            # auto permittivity detection
+            sim_orig = sim_data_orig.simulation
+            plane_eps = eps_fwd.monitor.geometry
+
+            # get permittivity without this structure
+            structs_no_struct = list(sim_orig.structures)
+            structs_no_struct.pop(structure_index)
+            sim_no_structure = sim_orig.updated_copy(structures=structs_no_struct)
+            eps_no_structure = sim_no_structure.epsilon(box=plane_eps, coord_key="centers")
+
+            # get permittivity without this structure
+            structs_inf_struct = list(sim_orig.structures)[structure_index + 1 :]
+            sim_inf_structure = sim_orig.updated_copy(
+                structures=structs_inf_struct,
+                medium=structure.medium,
+                monitors=[],
+            )
```

Comment:
Calling `updated_copy` on simulations always throws a red flag for me in terms of efficiency, because the simulation will get re-validated, which in some cases can be slow. In this particular case you're also doing this for every structure. I suspect this is going to get quite heavy e.g. in a metalens case where you have many structures so both re-validating each time may be costly, and it will be done many times (so at least O(N**2) scaling?)

Reply:
I think the answer here may be to refactor the `Simulation.epsilon` internals to have a function that can just get the epsilon of a list of structures without having to copy the simulation.

Reply:
Also computing over the entire simulation domain sounds like a giant overkill right?

Reply:
Interesting, thanks for the clarification!

> For things like metalens, it's likely to be a single structure since it's a geometry group. For example I tried this for both metalens and phc notebooks and it was negligible.

I guess that's what we try to teach people to use, but they don't always - but fair point! So the inside and outside eps will be computed only once?

> It would be worth also considering adding a validate: bool argument to updated_copy? I thought Weiliang implemented this somewhere recently.

You're right, this is currently on pre/2.8 only. Actually should we put this in 2.8 too just to avoid unexpected side-effects? It seems like there could be some since who knows what edge cases we haven't thought about.

> The epsilon is not computed over the entire simulation, just over the region defined by the .geometry of the graidient permittivity monitor.

Ah ok I see.

Reply:
> [For geometry group] So the inside and outside eps will be computed only once?

yes

> Actually should we put this in 2.8 too just to avoid unexpected side-effects?

I guess it's up to you, but that would be safer yea.  Maybe I should build in some logic to preferentially skip this calculation if the user still manually supplies `background_medium`.

Reply:
I thought that's what's gonna happen currently?

Yeah I think we should do 2.8 otherwise we're risking having to do another patch like few days later when we figure out some issue with this. :D

But wait for my merge before you change the base branch and cherry-pick this.

Reply:
ok, no rush